core/PRP/Geometry/plGBufferGroup.cpp

```diff
@@ -547,6 +560,113 @@ void plGBufferGroup::addIndices(const std::vector<unsigned short>& indices)
         fIdxBuffStorage[idx][i] = indices[i];
 }
 
+void plGBufferGroup::IPackIcicle(const plGeometrySpan* geoSpan, plIcicle* ice)
+{
+    size_t i;
+    for (i = 0; i < fIdxBuffStorage.size(); ++i)
+        if (geoSpan->getNumIndices() + fIdxBuffCounts[i] < kMaxNumIndicesPerBuffer)
+            break;
+    if (i == fIdxBuffStorage.size()) {
+        fIdxBuffStorage.push_back(nullptr);
+        fIdxBuffCounts.push_back(0);
+    }
+
+    unsigned int idxOffset = fIdxBuffCounts[i];
+    unsigned short* idxStorage = new unsigned short[idxOffset + geoSpan->getNumIndices()];
+    if (idxOffset != 0)
+        memcpy(idxStorage, fIdxBuffStorage[i], idxOffset * sizeof(unsigned short));
+    fIdxBuffCounts[i] += geoSpan->getNumIndices();
+    delete fIdxBuffStorage[i];
+    fIdxBuffStorage[i] = idxStorage;
+
+    for (size_t j = 0; j < geoSpan->getNumIndices(); ++j)
+        idxStorage[idxOffset + j] = geoSpan->getIndex(j) + (unsigned short)ice->getVStartIdx();
+
+    ice->setIBufferIdx(i);
+    ice->setIStartIdx(idxOffset);
+    ice->setILength(geoSpan->getNumIndices());
+}
+
+void plGBufferGroup::IPackVertexSpan(const plGeometrySpan* geoSpan, plVertexSpan* vSpan)
+{
+    size_t i;
+    for (i = 0; i < fVertBuffStorage.size(); ++i)
+        if (geoSpan->getNumVertices() + getNumVertices(i) < kMaxNumVertsPerBuffer)
+            break;
+    if (i == fVertBuffStorage.size()) {
+        fVertBuffStorage.push_back(nullptr);
+        fVertBuffSizes.push_back(0);
+        fCompGBuffStorage.push_back(nullptr);
+        fCompGBuffSizes.push_back(0);
+        fCells.emplace_back();
+    }
+
+    unsigned int vertOffset = fVertBuffSizes[i];
+    unsigned int vertBuffSize = vertOffset + (geoSpan->getNumVertices() * fStride);
+    unsigned char* vertStorage = new unsigned char[vertBuffSize];
+    if (vertOffset != 0)
+        memcpy(vertStorage, fVertBuffStorage[i], vertOffset);
+    fVertBuffSizes[i] = vertBuffSize;
+    delete fVertBuffStorage[i];
+    fVertBuffStorage[i] = vertStorage;
+
+    float* destp = (float*)(vertStorage + vertOffset);
+    for (const auto& vertex : geoSpan->getVertices()) {
+        *destp++ = vertex.fPosition.X;
+        *destp++ = vertex.fPosition.Y;
+        *destp++ = vertex.fPosition.Z;
```

Comment:
This type of type punning can generate very slow unaligned writes (yes, I know we already do that in other places :( )...  A memcpy might be better:

```c++
unsigned char *destp = vertStorage + vertOffset;
...
memcpy(destp, vertex.fPosition, sizeof(vertex.fPosition));
destp += sizeof(vertex.fPosition);
```

Reply:
Good point. Perhaps a pass over the old code would be relevant at some point.

Reply:
I was a little spooked by copying a class directly into POD, so I added a static_assert on the vector data being the expected size.

Reply:
Yep, the `static_assert`s are definitely reasonable here...